L1Trigger/L1THGCal/plugins/be_algorithms/HGCalSimCluster.cc

```diff
@@ -0,0 +1,256 @@
+#endif
+                // I want a map cell-> [ (pid, fraction),  ... 
+                std::unordered_map<uint32_t, std::vector<std::pair< uint64_t, float > > > simclusters;
+                for (auto& cluster : *sim_handle)
+                {
+                    auto pid= cluster.particleId(); // not pdgId
+                    const auto& hf = cluster.hits_and_fractions();
+                    for (const auto & p : hf ) 
+                    {
+                        simclusters[p.first].push_back( std::pair<uint64_t, float>( pid,p.second) ) ;
+                    }
+                }
+                
+#ifdef DEBUG
+                cout<<"[HGCalTriggerSimCluster]::[run] Run on digis"<<endl;
+#endif
+
+                //2. run on the digits,
+                for( const auto& digi : coll ) 
+                {
+                    DATA data;
+                    digi.decode(codec_,data);
+                    //2.A get the trigger-cell information energy/id
+                    const HGCTriggerDetId& moduleId = digi.getDetId<HGCTriggerDetId>(); // this is a module Det Id
+
+                    // there is a loss of generality here, due to the restriction imposed by the data formats
+                    // it will work if inside a module there is a data.payload with an ordered list of all the energies
+                    // one may think to add on top of it a wrapper if this stop to be the case for some of the data classes
+                    for(const auto& triggercell : data.payload)
+                    { 
+                            if(triggercell.hwPt()<=0) continue;
+
+                            const HGCalDetId tcellId(triggercell.detId());
+                            //uint32_t digiEnergy = data.payload; i
+                            auto digiEnergy=triggercell.p4().E();  
+                            double eta=triggercell.p4().Eta();
+                            double phi=triggercell.p4().Phi();
+                            //2.B get the HGCAL-base-cell associated to it / geometry
+                            //const auto& tc=geom->triggerCells()[ tcellId() ] ;//HGCalTriggerGeometry::TriggerCell&
+                            //for(const auto& cell : tc.components() )  // HGcell -- unsigned
+                            for(const auto& cell : geom->getCellsFromTriggerCell( tcellId()) )  // HGCcell -- unsigned
+                            {
+                                HGCalDetId cellId(cell);
+                                //2.C get the particleId and energy fractions
+                                const auto& particles =  simclusters[cellId]; // vector pid fractions
+                                for ( const auto& p: particles ) 
+                                {
+                                    const auto & pid= p.first;
+                                    const auto & fraction=p.second;
+                                    auto energy = fraction*digiEnergy;
```

Comment:
Thanks :-)
But here digiEnergy is the trigger cell energy and it seems like you add it to the cluster for each cell inside the trigger cell (the `addToCluster` call is done inside the cell loop). Naively I would think that the same energy is counted multiple times.